File: StructuralPatterns/Adapter/adapter.cpp
#include <iostream>

// Existing interface
class WeatherInCelsius {
public:
    virtual float getTemperature() {
        return 25.0; // returns temperature in Celsius
    }
};

// Third-party library (incompatible interface)
class WeatherInFahrenheit {
public:
    float fetchTemperature() {
        return 77.0; // returns temperature in Fahrenheit
    }
};

// Adapter class
class TemperatureAdapter : public WeatherInCelsius {
private:
    WeatherInFahrenheit* weatherInFahrenheit;
public:
    TemperatureAdapter(WeatherInFahrenheit* weather) : weatherInFahrenheit(weather) {}
    float getTemperature() override {
        float tempFahrenheit = weatherInFahrenheit->fetchTemperature();
        float tempCelsius = (tempFahrenheit - 32) * 5.0 / 9.0;
        return tempCelsius; // Converts Fahrenheit to Celsius
    }
};

// Client code
void displayTemperature(WeatherInCelsius* weatherSource) {
    std::cout << "The current temperature is " << weatherSource->getTemperature() << "°C." << std::endl;
}

int main() {
    // Using the existing interface
    WeatherInCelsius celsiusWeather;
    displayTemperature(&celsiusWeather);

    // Using the third-party library through the adapter
    WeatherInFahrenheit fahrenheitWeather;
    TemperatureAdapter adapter(&fahrenheitWeather);
    displayTemperature(&adapter);

    return 0;
}

File: StructuralPatterns/Adapter/adapter.py
# In this example, TemperatureAdapter is serving as the bridge between WeatherInCelsius and WeatherInFahrenheit.
# It adapts the interface of WeatherInFahrenheit to match that of WeatherInCelsius, allowing them to work together seamlessly.
# Existing interface
class WeatherInCelsius:
    def get_temperature(self) -> float:
        return 25.0  # returns temperature in Celsius


# Third-party library with incompatible interface
class WeatherInFahrenheit:
    def fetch_temperature(self) -> float:
        return 77.0  # returns temperature in Fahrenheit


# Adapter class
class TemperatureAdapter:
    def __init__(self, weather_in_fahrenheit: WeatherInFahrenheit) -> None:
        self.weather_in_fahrenheit = weather_in_fahrenheit

    def get_temperature(self) -> float:
        temp_fahrenheit = self.weather_in_fahrenheit.fetch_temperature()
        temp_celsius = (temp_fahrenheit - 32) * 5 / 9
        return temp_celsius  # converts Fahrenheit to Celsius


# Client code
def display_temperature(weather_source):
    print(f"The current temperature is {weather_source.get_temperature()}°C.")


if __name__ == "__main__":
    # Using the existing interface
    celsius_weather = WeatherInCelsius()
    display_temperature(celsius_weather)

    # Using the third-party library through the adapter
    fahrenheit_weather = WeatherInFahrenheit()
    adapter = TemperatureAdapter(fahrenheit_weather)
    display_temperature(adapter)
